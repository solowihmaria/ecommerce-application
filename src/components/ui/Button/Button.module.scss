.button {
    cursor: pointer;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 0.75rem 1.5rem;
    border: 1px solid transparent;
    border-radius: 0.25rem;
    font-family: inherit;
    font-size: 1rem;
    font-weight: 500;
    line-height: 1.5;
    transition: all 0.2s ease;

    &:disabled {
        cursor: not-allowed;
        color: #a0a0a0;
        opacity: 0.5;
        background-color: #e0e0e0;
    }

    &.primary {
        border-color: #000;
        color: #fff;
        background-color: #000;

        &:hover:not(:disabled) {
            background-color: #333;
        }
    }

    &.secondary {
        border-color: #e0e0e0;
        color: #000;
        background-color: #f5f5f5;

        &:hover:not(:disabled) {
            background-color: #e8e8e8;
        }
    }

    &.danger {
        border-color: #ff4d4f;
        color: #ff4d4f;
        background-color: #fff;

        &:hover:not(:disabled) {
            background-color: #fff2f2;
        }
    }

    &.outline {
        border-color: #000;
        color: #000;
        background-color: transparent;

        &:hover:not(:disabled) {
            background-color: rgb(0 0 0 / 5%);
        }
    }

    &.ghost {
        padding: 0.5rem 1rem;
        border-color: transparent;
        color: #000;
        background-color: transparent;

        &:hover:not(:disabled) {
            text-decoration: underline;
            background-color: rgb(0 0 0 / 5%);
        }
    }
}

.spinner {
    display: inline-block;
    width: 1rem;
    height: 1rem;
    margin-left: 0.75rem;
    border: 2px solid rgb(255 255 255 / 30%);
    border-top-color: currentcolor;
    border-radius: 50%;
    animation: spin 0.8s linear infinite;

    .secondary &,
    .outline &,
    .ghost & {
        border-top-color: #000;
    }

    .danger & {
        border-top-color: #ff4d4f;
    }
}

@keyframes spin {
    to {
        transform: rotate(360deg);
    }
}

.button.loading {
    cursor: wait;

    &.primary .spinner {
        border-top-color: white;
    }
}
