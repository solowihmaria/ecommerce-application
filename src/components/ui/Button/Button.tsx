import React from 'react';
import styles from './Button.module.scss';
import clsx from 'clsx';
import type { ButtonProps } from './Button.types';

export const Button = ({
    variant = 'primary',
    loading = false,
    className = '',
    children,
    ...props
}: ButtonProps & { loading?: boolean }) => {
    return (
        <button
            className={clsx(styles.button, styles[variant], className, {
                [styles.disabled]: props.disabled,
                [styles.loading]: loading,
            })}
            disabled={loading || props.disabled}
            {...props}
        >
            {children}
            {loading && <span className={styles.spinner} />}
        </button>
    );
};
