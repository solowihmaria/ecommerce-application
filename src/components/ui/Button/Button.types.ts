import type { ButtonHTMLAttributes } from 'react';

export type ButtonVariant =
    | 'primary'
    | 'secondary'
    | 'danger'
    | 'outline'
    | 'ghost';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    variant?: ButtonVariant;
}
