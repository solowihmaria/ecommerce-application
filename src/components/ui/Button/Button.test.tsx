import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Button } from './Button';
import type { ButtonProps } from './Button.types';

describe('<Button/> component', () => {
    test('Render Button component - without any properties', () => {
        render(<Button>button name</Button>);
        const button = screen.getByRole('button');

        expect(button).toBeInTheDocument();
        expect(button).toHaveClass('button');
        expect(button).toHaveClass('primary');
        expect(button).toMatchSnapshot();
    });

    test('Render Button component with properties', () => {
        const properties: ButtonProps = {
            type: 'button',
            className: 'testClass',
            onClick: () => {},
            disabled: false,
            variant: 'secondary',
            children: 'button name',
        };
        render(<Button {...properties} />);
        const button = screen.getByRole('button');

        expect(button).toHaveClass('testClass');
        expect(button).toHaveClass('secondary');
        expect(button).toMatchSnapshot();
    });

    test('Render Button component - disabled', () => {
        const properties: ButtonProps = {
            type: 'button',
            className: 'testClass',
            onClick: () => {},
            disabled: true,
            variant: 'secondary',
            children: 'button name',
        };
        render(<Button {...properties} />);
        const button = screen.getByRole('button');

        expect(button).toBeInTheDocument();
        expect(button).toBeDisabled();
        expect(button).toMatchSnapshot();
    });

    test('Render Button component - check that handler is triggered on click', () => {
        const onClick = jest.fn();

        render(<Button onClick={onClick}>button name</Button>);

        const button = screen.getByRole('button');
        fireEvent.click(button);
        expect(onClick).toHaveBeenCalledTimes(1);
    });
});
