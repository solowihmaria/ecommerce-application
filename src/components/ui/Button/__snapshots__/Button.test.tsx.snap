// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> component Render Button component - disabled 1`] = `
<button
  class="button secondary testClass disabled"
  disabled=""
  type="button"
>
  button name
</button>
`;

exports[`<Button/> component Render Button component - without any properties 1`] = `
<button
  class="button primary"
  type="button"
>
  button name
</button>
`;

exports[`<Button/> component Render Button component with properties 1`] = `
<button
  class="button secondary testClass"
  type="button"
>
  button name
</button>
`;
